src/Node/HTTP2.purs

```diff
@@ -0,0 +1,313 @@
+-- | will complete next, but we don’t know which one. We want to wait for
+-- | all of them at the same time.
+-- |
+-- | ```
+-- | parOneOf
+-- |   [ event1
+-- |   , event2
+-- |   , event3
+-- |   ]
+-- | ```
+module Node.HTTP2
+  ( HeadersObject(..)
+  , toHeaders
+  , sensitiveHeaders
+  , headerKeys
+  , headerString
+  , headerArray
+  , headerStatus
+  , OptionsObject(..)
+  , toOptions
+  , Flags
+  , SettingsObject(..)
+  ) where
+
+import Prelude
+
+import Control.Monad.Except (runExcept)
+import Data.Either (hush)
+import Data.Maybe (Maybe, fromJust)
+import Data.Newtype (class Newtype)
+import Data.Traversable (traverse)
+import Foreign (Foreign, readArray, readInt, readString, unsafeToForeign)
+import Foreign.Index (readProp)
+import Foreign.Keys (keys)
+import Foreign.Object (union)
+import Partial.Unsafe (unsafePartial)
+import Unsafe.Coerce (unsafeCoerce)
+
+-- | An HTTP/2 “headers object.”
+-- |
+-- | https://nodejs.org/docs/latest/api/http2.html#headers-object
+-- |
+-- | Construct with the `toHeaders` function.
+-- | The “no headers” literal is `toHeaders {}`.
+-- |
+-- | The `Monoid` instance allows us to merge `HeadersObject` with `<>`.
+newtype HeadersObject = HeadersObject Foreign
+
+derive instance Newtype HeadersObject _
+
+instance Semigroup HeadersObject where
+  append l r = unionHeadersImpl l r
+
+instance Monoid HeadersObject where
+  mempty = unsafeCoerce {}
+
+foreign import unionHeadersImpl :: HeadersObject -> HeadersObject -> HeadersObject
+
+-- | Use this function to construct a `HeadersObject` object.
+-- |
+-- | Rules for Headers:
+-- |
+-- | https://nodejs.org/docs/latest/api/http2.html#headers-object
+-- |
+-- | > Headers are represented as own-properties on JavaScript objects.
+-- | > The property keys will be serialized to lower-case.
+-- | > Property values should be strings (if they are not they will
+-- | > be coerced to strings) or an Array of strings (in order to send
+-- | > more than one value per header field).
+-- |
+-- | This function provides no type-level enforcement of these rules.
+-- |
+-- | Example:
+-- |
+-- | ```
+-- | toHeaders
+-- |   { ":status": "200"
+-- |   , "content-type": "text-plain"
+-- |   , "ABC": ["has", "more", "than", "one", "value"]
+-- |   }
+-- | ```
+toHeaders :: forall r. Record r -> HeadersObject
+toHeaders = HeadersObject <<< unsafeToForeign
+
+-- | Use this function to construct a “sensitive” `HeadersObject`.
+-- |
+-- | https://nodejs.org/docs/latest/api/http2.html#sensitive-headers
+-- |
+-- | Example:
+-- |
+-- | ```
+-- | toHeaders
+-- |   { "content-type": "text-plain"
+-- |   }
+-- | <>
+-- | sensitiveHeaders
+-- |   { "cookie": "some-cookie"
+-- |   , "other-sensitive-header": "very secret data"
+-- |   }
+-- | ```
+foreign import sensitiveHeaders :: forall r. Record r -> HeadersObject
+
+-- | Get all of the keys from a `HeadersObject`.
+-- |
+-- | The value pointed to by each key may be either a `String`
+-- | or an `Array String`.
```

Comment:
Would this be better represented as an `Object HeaderValue` if we know it's always an object (this lets you avoid things like your record coercion for the monoid instance), paired with functions for operating on a value such like `caseHeaderString` and `caseHeaderArray` a la

https://github.com/purescript-contrib/purescript-argonaut-core/blob/v7.0.0/src/Data/Argonaut/Core.purs#L107-L107

Reply:
Yeah, let's talk about this.

Your scheme would account for the cases where the value can be a `String` or an `Array String`.

https://nodejs.org/docs/latest/api/http2.html#headers-object

But what about __Sensitive headers__, where the __*key*__ is not a `String`, rather some array of who even knows what?

https://nodejs.org/docs/latest/api/http2.html#sensitive-headers

It's so difficult to provide for a complete accounting of all the corner cases of the Node.js API that I thought it would be best to defer the problem by presenting the headers as a `Foreign`. Anyone who wants to write a server package on top of this one can present an API with an `Object HeaderValue`, and then `unsafeToForeign` to call into this API.

And same for the `OptionsObject`. It’s also a `Foreign`, so anyone can write their own `Options HTTP2Option` for it.

Reply:
Yep, that makes sense then — although it makes your keys implementation a bit dodgy, as it would throw on an array?

Reply:
> keys implementation a bit dodgy, as it would throw on an array?

Uh, yes, probably. (I think normally you would not be reading from sensitive headers, only writing to them?) We could use the `Foreign` module functions to improve the implementation so that it would work in the case of reading sensitive headers.

Also currently `<>` for two sensitive `HeadersObject`s is probably broken.